src/Symfony/Component/DependencyInjection/ContainerBuilder.php

```diff
@@ -435,12 +431,24 @@ public function get($id, $invalidBehavior = ContainerInterface::EXCEPTION_ON_INV
 
         try {
             $definition = $this->getDefinition($id);
+            if ($definition instanceof ServiceAwareDefinition) {
+                return $definition->getService();
+            }
         } catch (ServiceNotFoundException $e) {
-            if (ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE !== $invalidBehavior) {
-                return;
+            if ($this->compiled && ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE === $invalidBehavior) {
```

Comment:
For BC, however i tend to debug unknown id's before state. So imo the whole try/catch can also be removed...